### PR TITLE
test: Add Writing Flow tests

### PR DESCRIPTION
**Pull request by dcalhoun**
## What?

Add automated test coverage for crucial writing flow interactions within the
native mobile editor.

## Why?

Automated test coverage of these important features will enable us to avoid
introducing regressions while we add new features and refactor existing code to
optimize performance.

## How?

Leverage existing integration test helpers to cover portions of the mobile app 
writing flow with automated tests. 

## Testing Instructions

n/a

### Testing Instructions for Keyboard

n/a

## Screenshots or screencast 
n/a

### CODE REVIEW

File: test/native/integration/editor-history.native.js
Comment (dcalhoun) [inline test/native/integration/editor-history.native.js:1]: @fluiddot what do you think of this file placement and approach? There are subjects that are difficult to colocate with a unit of code, i.e editor history applies to various blocks. 

If we like this approach, we could also hoist some of the text formatting tests set up for Paragraph to this location and run them against various Rich Text blocks, e.g. Heading, Quote.

File: packages/block-library/src/preformatted/test/edit.native.js [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
+ * WordPress dependencies
+ */
+import { ENTER } from '@wordpress/keycodes';
 
 /**
  * Internal dependencies
  */
 import PreformattedEdit from '../edit';
 
-describe( 'core/more/edit/native', () => {
```
Comment (dcalhoun) [inline packages/block-library/src/preformatted/test/edit.native.js:11]: Even after reviewing the project history, I am not sure where the `core/more/edit/native` name originated. It seemed incorrect and misleading. "Preformatted" felt clearer.

File: packages/block-library/src/pullquote/test/edit.native.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+			selectionStart: string.length,
+			selectionEnd: string.length,
+		} );
+		fireEvent( pullquoteTextInput, 'onKeyDown', {
+			nativeEvent: {},
+			preventDefault() {},
+			keyCode: ENTER,
+		} );
+
+		// TODO: Determine a way to type after pressing ENTER within the block.
```
Comment (dcalhoun) [inline packages/block-library/src/pullquote/test/edit.native.js:49]: Similar comments can be found through this PR. The `changeTextOfRichText` and `changeAndSelectTextOfRichText` helpers do not yet support typing multiple times into the same Rich Text input. Attempts to do so will clear out previously "typed" text. 

I explored adding this support, but decided to consider it out of scope of this PR as it was taking too long to implement. I explored tracking previous text entries in local scope, further mocking Aztec implementations, and subscribing to store events. Unfortunately, I did not arrive to a working solution. More research is needed.

File: test/native/integration/editor-history.native.js
Comment (fluiddot) [reply test/native/integration/editor-history.native.js:1]: Definitely, it's tricky to find the right place for some of the test cases, especially if they are not tied to a component or block. This location looks good to me, in fact, replicates the [`integration` folder](https://github.com/WordPress/gutenberg/tree/trunk/test/integration) we have for the web version 👍 .

**Comment by fluiddot** [comment]
> ## Why?
> Automated test coverage of these important features will enable us to avoid introducing regressions while we add new features and refactor existing code to optimize performance.

@dcalhoun @derekblank I understand that we'll be able to spot most of the regressions with these tests, but I'd like to note that part of the logic of the writing flow relies on the Aztec native library, which won't be covered with integration tests. For covering Aztec we'd need to do it via E2E tests. Do you foresee this caveat would impact the confidence in refactoring existing code?

**Comment by dcalhoun** [comment]
> I understand that we'll be able to spot most of the regressions with these tests, but I'd like to note that part of the logic of the writing flow relies on the Aztec native library, which won't be covered with integration tests. For covering Aztec we'd need to do it via E2E tests. Do you foresee this caveat would impact the confidence in refactoring existing code?

@fluiddot good question. I think we should consider this question in two contexts: refactoring to replace Aztec and refactoring for performance optimizations. 

For the former, I perceive the question will largely become moot as Aztec would no longer be a requirement. It would, in theory, be replaced by React Native's `TextInput`. We could rely upon e2e tests to verify the relationship and communication with `TextInput` functions as expected, but we should likely "draw a line" somewhere in regards to what we test. The third-party code is covered by its own tests. IMO us testing third-party code likely does not deliver value worth the effort. 

For the latter, your question is very relevant. By relying upon integration tests that mock Aztec, we expose ourselves to regressions that relate to Aztec itself or communication with it. I imagine Aztec itself is/should be covered with its own tests, which means we only need to decide if we want to test our communication with Aztec. This communication is likely pretty important and I could understand an argument we should test it. 

With these integration tests, I decided to balance speed, maintainability, coverage, and long-term goals. I believe they will provide value for one day replacing Aztec. IMO, this is the biggest and most necessary "win." I believe they provide _some_ value for performance optimizations we choose to make. I imagine we will still perform manual tests for any performance optimizations we make, both due to the lack of Aztec coverage you reference and the fact that testing/verifying performance optimizations often benefit from experiencing the changes in "real world" scenarios. 

I will also note that we do [plan to add e2e tests](https://docs.google.com/spreadsheets/d/1YW6Nu1hZwc9yBkT8QaEkKKS1z0WJmQpQ-XHO2VevqWw/edit#gid=0&range=A59:J81) for aspects that are not easy or possible to test with integration tests, e.g. keyboard swiping. 

All that said, my thought was that integration tests for the writing flow made sense. However, I am open to approaching this in a different way with e2e tests. WDYT?

**Comment by fluiddot** [comment]
> For the former, I perceive the question will largely become moot as Aztec would no longer be a requirement. It would, in theory, be replaced by React Native's `TextInput`. We could rely upon e2e tests to verify the relationship and communication with `TextInput` functions as expected, but we should likely "draw a line" somewhere in regards to what we test. The third-party code is covered by its own tests. IMO us testing third-party code likely does not deliver value worth the effort.

Good point. I agree that we should rely as much as we can on integration tests. The rest of the cases that can't be done with integration tests could be left for the E2E tests, like for example, any test that needs visual testing. In this regard, I'm wondering how many of the writing flow tests would need it 🤔.

> For the latter, your question is very relevant. By relying upon integration tests that mock Aztec, we expose ourselves to regressions that relate to Aztec itself or communication with it. I imagine Aztec itself is/should be covered with its own tests, which means we only need to decide if we want to test our communication with Aztec. This communication is likely pretty important and I could understand an argument we should test it.

I understand that Aztec already has its tests to cover the logic. However, [as we have a custom implementation of Aztec for Gutenberg](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-aztec/README.md), I don't think that part is actually covered. Eventually, if we consider adding tests for this area, I wouldn't add tests to cover all the functionality but focus on the most critical parts like the communication with the editor as you mentioned.

> With these integration tests, I decided to balance speed, maintainability, coverage, and long-term goals. I believe they will provide value for one day replacing Aztec. IMO, this is the biggest and most necessary "win." I believe they provide _some_ value for performance optimizations we choose to make. I imagine we will still perform manual tests for any performance optimizations we make, both due to the lack of Aztec coverage you reference and the fact that testing/verifying performance optimizations often benefit from experiencing the changes in "real world" scenarios.

Makes total sense, as I mentioned earlier, it would be great to rely on integration tests as much as we can due to all the benefits they provide. In terms of performance, I understand that we can verify parts related to React like re-renders or render time with integration tests. But if we'd like to cover all performance areas, we still would need to have E2E tests (probably in the host apps) to measure performance as close to the user experience.

> I will also note that we do [plan to add e2e tests](https://docs.google.com/spreadsheets/d/1YW6Nu1hZwc9yBkT8QaEkKKS1z0WJmQpQ-XHO2VevqWw/edit#gid=0&range=A59:J81) for aspects that are not easy or possible to test with integration tests, e.g. keyboard swiping.

That's great, thanks for sharing 🙇 !

> All that said, my thought was that integration tests for the writing flow made sense. However, I am open to approaching this in a different way with e2e tests. WDYT?

@dcalhoun These integration tests make total sense to me. My comment was just along the lines of making us aware that we won't fully cover all the logic related to the writing flow. I'd advocate following this approach and covering the native parts with future E2E tests for the most brittle areas like the communication with Aztec 👍.

**Comment by dcalhoun** [comment]
> The rest of the cases that can't be done with integration tests could be left for the E2E tests, like for example, any test that needs visual testing. In this regard, I'm wondering how many of the writing flow tests would need it 🤔.

The [test case spreadsheet](https://docs.google.com/spreadsheets/d/1YW6Nu1hZwc9yBkT8QaEkKKS1z0WJmQpQ-XHO2VevqWw/edit#gid=0&range=A59:J81) represents my assessment of which type of test each test case needs. We, of course, can always reassess that, and I welcome other weighing in on that subject. Inline comments on the spreadsheet is likely the most productive form of discussion. 

> However, [as we have a custom implementation of Aztec for Gutenberg](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-aztec/README.md), I don't think that part is actually covered.

Good point. 

> In terms of performance, I understand that we can verify parts related to React like re-renders or render time with integration tests. But if we'd like to cover all performance areas, we still would need to have E2E tests (probably in the host apps) to measure performance as close to the user experience.

Agreed. It is worth noting that running automated e2e performance tests would likely require leveraging a physical device farm to achieve semi-realistic results, i.e. I am not sure performance tests run on emulators would provide much value. 


> My comment was just along the lines of making us aware that we won't fully cover all the logic related to the writing flow. I'd advocate following this approach and covering the native parts with future E2E tests for the most brittle areas like the communication with Aztec 👍.

@fluiddot thank you for sharing your perspective. Much appreciated.

**Comment by fluiddot** [comment]
> The [test case spreadsheet](https://docs.google.com/spreadsheets/d/1YW6Nu1hZwc9yBkT8QaEkKKS1z0WJmQpQ-XHO2VevqWw/edit#gid=0&range=A59:J81) represents my assessment of which type of test each test case needs. We, of course, can always reassess that, and I welcome other weighing in on that subject. Inline comments on the spreadsheet is likely the most productive form of discussion.

Great, I'll take a look and share feedback 👍 .

**Review by derekblank — APPROVED** [review]
These look great, thank you for adding these test cases. Approved. 🚀

**Pull request merged**